### DISCUSSION TITLE
Distributing HOOMD component on conda-forge

### DISCUSSION THREAD

**Discussion by mphoward** [glotzerlab/hoomd-blue#1935] [Support] [answered]
Hi All,

This is continuation of the discussion in https://github.com/glotzerlab/hoomd-blue/discussions/1930#discussioncomment-11219230.

We have a HOOMD component  [azplugins](https://github.com/mphowardlab/azplugins) that mostly follows the [hoomd-component-template](https://github.com/glotzerlab/hoomd-component-template). With this new template, azplugins gets installed into the same location as HOOMD if you follow the [build instructions](https://hoomd-blue.readthedocs.io/en/v4.9.1/components.html#building-an-external-component).

My question is: will needing to install azplugins into hoomd cause issues if we wanted to distribute azplugins on conda-forge? I am not super familiar with conda-forge recipes beyond basic PyPI mirrors, but it seems like it could be problematic to install a package into another package. If yes, one way to get around this would be to re-enable "external" builds of components (like were present in prior versions of HOOMD) so that add-on components are self-contained. Is there anything in the current design of HOOMD that would prevent us from modifying the build to support installing a component outside of the HOOMD package?

Thanks for your help!

Regards,
Mike

**Comment by joaander** [comment, marked as answer]
> My question is: will needing to install azplugins into hoomd cause issues if we wanted to distribute azplugins on conda-forge? I am not super familiar with conda-forge recipes beyond basic PyPI mirrors, but it seems like it could be problematic to install a package into another package.

A conda-forge package is a compressed tar file of all the files it installs. packages are not given exclusive ownership over any single directory in the environment. `hoomd-component-template` installs components in to subdirectories under `site-packages/hoomd`. As long a component does not overwrite any of `hoomd`'s files, there will be no problems.

> If yes, one way to get around this would be to re-enable "external" builds of components (like were present in prior versions of HOOMD) so that add-on components are self-contained. Is there anything in the current design of HOOMD that would prevent us from modifying the build to support installing a component outside of the HOOMD package?

You would need to adjust the RPATH so that the `.so` files created by the component find the `_hoomd`, `_md`, etc... `.so` files that they link to.

https://github.com/glotzerlab/hoomd-component-template/blob/d97980ec5997b835ff83640447d40fcc22a18fb1/src/CMakeLists.txt#L27-L31

It is the choice of the component author whether they would prefer their python package to be included under `site-packages/hoomd/{component-name}` or at the `site-packages/{component-name}` level.

**Reply by mphoward** [reply]
OK thanks! I am happy to have azplugins install under hoomd if this is not an issue. It will be simpler to maintain long term. We'll look into this, and I'll attach a link to the recipe for others' reference if we get it to work.
